### PR TITLE
Convert examples from axy to test cases part 4

### PR DESCRIPTION
**Pull request by Nicoretti**
- **Add formatter tests**
- **Add transaction test marker**
- **Add transaction tests**
- **Add test marker for exception tests**
- **Add tests for exceptions**

### CODE REVIEW

File: test/integration/exceptions_test.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+@pytest.mark.exceptions
+def test_duplicated_select_of_column(connection):
+    query = "SELECT user_name, user_name FROM USERS;"
+    with pytest.raises(ExaRuntimeError):
+        connection.execute(query)
+
+
+@pytest.mark.exceptions
+def test_attempt_to_Run_query_on_closed_connection(connection):
```
Comment (tkilias) [inline test/integration/exceptions_test.py:74]:
```suggestion
def test_attempt_to_run_query_on_closed_connection(connection):
```

File: test/integration/exceptions_test.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        ),
+        pyexasol.connect(
+            dsn=dsn,
+            user=user,
+            password=password,
+            schema=schema,
+        )
+    ]
+
+    connections[1].execute(f"KILL SESSION {connections[0].session_id()}")
```
Comment (tkilias) [inline test/integration/exceptions_test.py:97]: connections[1] doesn't get closed by the test

File: test/integration/formatter_test.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    params = {"value": value}
+    query = "SELECT {value};"
+    actual = connection.execute(query, params).fetchval()
+
+    assert expected == actual
+
+
+@pytest.mark.format
+def test_safe_quote_conversion(connection, schema):
+    params = {"table": (schema, "USERS")}
```
Comment (tkilias) [inline test/integration/formatter_test.py:21]: here a table with a lower case table name would be good, to check if the quoting actually works

File: test/integration/formatter_test.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        'Zachary Thompson'
+    ]
+    actual = connection.execute(query, params).fetchcol()
+
+    assert expected == actual
+
+
+@pytest.mark.format
+def test_safe_identifier_conversion(connection):
+    params = {"table": "USERS"}
```
Comment (tkilias) [inline test/integration/formatter_test.py:65]: Probably similar as for the quoting, a lowercase table name would be good

File: test/integration/formatter_test.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+@pytest.mark.format
+@pytest.mark.parametrize("value", [
+    True, 'True', "TRUE", 'true'
+])
+def test_string_conversion(connection, value):
+    params = {"is_female": value}
+    query = cleandoc("""
+        SELECT user_name FROM USERS
+        WHERE is_female IS {is_female!r}
```
Comment (tkilias) [inline test/integration/formatter_test.py:47]: I would almost say, we should actually return the formatted value

File: test/integration/formatter_test.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+def test_safe_decimal_conversion(connection, value, expected):
+    params = {"value": value}
+    query = "SELECT {value!d};"
+    result = connection.execute(query, params)
+    actual = result.fetchval()
+
+    assert expected == actual
+
+
+@ pytest.mark.format
```
Comment (tkilias) [inline test/integration/formatter_test.py:112]:
```suggestion
@pytest.mark.format
```

File: test/integration/transaction_test.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import pytest
+import pyexasol
+
+
+# For the fetch_dict tests we need to configure the connection accordingly (fetch_dict=True)
```
Comment (tkilias) [inline test/integration/transaction_test.py:5]: would it makes sense to add this as docstring

File: test/integration/transaction_test.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+@pytest.fixture
+def connection(dsn, user, password, schema):
+    con = pyexasol.connect(
+        dsn=dsn,
+        user=user,
+        password=password,
+        schema=schema,
+        autocommit=False
```
Comment (tkilias) [inline test/integration/transaction_test.py:13]: I don't see fetch_dict=True

**Review by tkilias — CHANGES_REQUESTED** [review]

File: test/integration/transaction_test.py (hunk 8 above)
Comment (Nicoretti) [reply test/integration/transaction_test.py:13]: Not needed in this test.

File: test/integration/transaction_test.py (hunk 7 above)
Comment (Nicoretti) [reply test/integration/transaction_test.py:5]: No but I see the comment is wrong ... thanks :+1:

File: test/integration/formatter_test.py (hunk 4 above)
Comment (Nicoretti) [reply test/integration/formatter_test.py:65]: works with lower case -> changed.

File: test/integration/formatter_test.py (hunk 3 above)
Comment (Nicoretti) [reply test/integration/formatter_test.py:21]: This does not work with lower case table name.

**Review by tkilias — APPROVED** [review]

**Pull request merged**